web/tests/stream_data.test.js

```diff
@@ -166,16 +162,21 @@ test("basics", () => {
     assert.ok(!stream_data.is_default_stream_id(social.stream_id));
     assert.ok(!stream_data.is_default_stream_id(999999));
 
-    assert.equal(stream_data.slug_to_name("2-social"), "social");
-    assert.equal(stream_data.slug_to_name("2-whatever"), "social");
-    assert.equal(stream_data.slug_to_name("2"), "social");
-
+    // "new" correct url formats
+    assert.equal(stream_data.slug_to_stream_id("2-social"), 2);
+    assert.equal(stream_data.slug_to_stream_id("2"), 2);
+    // we still get 2 because it's a valid stream id
+    assert.equal(stream_data.slug_to_stream_id("2-whatever"), 2);
+    // invalid stream id
+    assert.equal(stream_data.slug_to_stream_id("999-social"), undefined);
     // legacy
-    assert.equal(stream_data.slug_to_name("25-or-6-to-4"), "25-or-6-to-4");
-    assert.equal(stream_data.slug_to_name("2something"), "2something");
+    assert.equal(stream_data.slug_to_stream_id("social"), 2);
 
-    assert.equal(stream_data.slug_to_name("99-whatever"), "99-whatever");
-    assert.equal(stream_data.slug_to_name("99whatever"), "99whatever");
+    // invalid formats
+    assert.equal(stream_data.slug_to_stream_id("25-or-6-to-4"), undefined);
+    assert.equal(stream_data.slug_to_stream_id("2something"), undefined);
+    assert.equal(stream_data.slug_to_stream_id("99-whatever"), undefined);
+    assert.equal(stream_data.slug_to_stream_id("99whatever"), undefined);
```

Comment:
The first and third of these are validly formatted in the new style: they mean a stream/channel with ID 25 or 99 respectively.

All four are validly formatted in the old style: they mean streams with those names.

Reply:
Currently `slug_to_stream_id` just gets a valid stream id, and `decode_operand` is doing the rest of the work, including returning the full operand (i.e. the name) if there's no valid stream id found.

`25-or-6-to-4` and `99-whatever` returned `undefined` because `25` and `9` weren't valid ids. I could do a followup test adding streams with those ids and show that it returns the id then?

I'm thinking maybe I should keep `slug_to_name` and these tests cases, and add a new function `slug_to_id` that is used by `slug_to_name`.

Reply:
> returning the full operand (i.e. the name) if there's no valid stream id found.

I see. So I guess then the behavior question comes down to: if the stream operand we're looking at
* doesn't correspond to any known stream using either format,
* but does have the shape consistent with the post-2018 format (i.e. it matches `/^(\d+)(?:-.*)?$/`),
* then should we interpret it according to the post-2018 format or the pre-2018 format?

I think the preferable answer is we should interpret it with the post-2018 format. It seems like it'll be way more likely that an operand `99-whatever` is a post-2018 link to an (unknown) stream with ID 99 than that it's a pre-2018 link to an (unknown) stream with name "99-whatever". Both because 2018 was a long time ago now, and because it'd be kind of a coincidence for the stream name to have the shape `/^(\d+)(?:-.*)?$/` anyway.

The behavior in this version is effectively that we interpret such an operand with the pre-2018 format — that's what it means to take the whole operand as the stream name, instead of taking the leading `(\d+)` group as the stream ID.

(Perhaps relatedly, I'd say 25 and 9 are always "valid stream IDs", regardless of what streams the client may or may not know about, or what streams may or may not exist on the server.)

Reply:
One bonus of that approach is that it means you can have `decode_operand` always return (a stringification of) a numeric stream ID when the operand is `stream`/`channel`. Then everything that consumes such operands can know that's the case, and not have to ever wonder if it's a stream name rather than a stream ID.

Reply:
> I'm thinking maybe I should keep `slug_to_name` and these tests cases, and add a new function `slug_to_id` that is used by `slug_to_name`.

I tried this and liked it, and will push that shortly.

If we're just showing "unknown channel" without more copy, then I don't think it matters what we return when we can't identify a stream id. Does that sound right?

Reply:
> If we're just showing "unknown channel" without more copy, then I don't think it matters what we return when we can't identify a stream id. Does that sound right?

Yeah, it might not matter in that case, depending on what else the code that consumes these results is doing with the information. (I haven't looked.)

Reply:
If it truly doesn't matter, though, it'd be best to not propagate all the mess of the full slug — better to either return the presumed stream ID, or some fixed value like "" or null or undefined.

Passing around the full slug when it doesn't mean anything will be an invitation for bugs, as future people working on other code assume it's supposed to potentially mean something and propagate new complicated logic trying to interpret it.

Reply:
>  depending on what else the code that consumes these results is doing with the information. (I haven't looked.)

Looking -- `slug_to_name` is only called by `decode_operand` which is only called by `parse_narrow`, which is called in two places:

*`hashchange.do_hashchange_normal`, and the parsed terms are sent to `message_view.show`
  * I haven't looked at this in depth, but it seems to be the stuff we've been focusing on so far, which is the navbar header and the message feed contents
* `hash_util.decode_stream_topic_from_url`, only used in `copy_and_paste.try_stream_topic_syntax_text`
  * Here it makes sense to return `null` from `decode_stream_topic_from_url` if we can't find a valid stream id, so I'll change that but it doesn't make a difference for this conversation.

I double checked that what we return doesn't seem to affect normal channel narrowing or search bar or reply behavior or anything, so I think we can return whatever we want here.

So here's what we could do:

* Return a valid stream id, if we can find one from a prepended number on the slug (new format) or from the full slug as a valid stream name (old format)
* return an empty string (or undefined?) from `slug_to_name` (and therefore `decode_operand`) if we can't find a valid stream id

If we did this, it would be pretty much what I had before, so I'd go back to not having `slug_to_name` and only `slug_to_stream_id` which can return `undefined`, and have `decode_operand` call that directly. Probably the main difference would be defaulting to an empty string here, instead of falling back to returning the `operand`:

```js
    if (util.canonicalize_stream_synonyms(operator) === "stream") {
        return stream_data.slug_to_id(operand) ?? "";
    }
```

Reply:
I'm going to do this